File: src/main/java/net/ballmerlabs/scatterbrainsdk/HandshakeResult.java
package net.ballmerlabs.scatterbrainsdk;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;

public class HandshakeResult implements Parcelable {


    public enum TransactionStatus {
        STATUS_SUCCESS,
        STATUS_FAIL
    }
    
    protected HandshakeResult(Parcel in) {
        identities = in.readInt();
        messages = in.readInt();
        status = TransactionStatus.values()[in.readInt()];
    }

    public static final Creator<HandshakeResult> CREATOR = new Creator<HandshakeResult>() {
        @Override
        public HandshakeResult createFromParcel(Parcel in) {
            return new HandshakeResult(in);
        }

        @Override
        public HandshakeResult[] newArray(int size) {
            return new HandshakeResult[size];
        }
    };

    public final int identities;
    public final int messages;
    public final TransactionStatus status;

    public HandshakeResult(
            int identities,
            int messages,
            TransactionStatus status
    ) {
        this.status = status;
        this.messages = messages;
        this.identities = identities;
    }

    @NonNull
    public HandshakeResult from(HandshakeResult stats) {
        final TransactionStatus status;
        if (stats.status == TransactionStatus.STATUS_FAIL ||
                this.status == TransactionStatus.STATUS_FAIL) {
            status = TransactionStatus.STATUS_FAIL;
        } else {
            status = stats.status;
        }
        return new HandshakeResult(
                stats.identities + this.identities,
                stats.messages + this.messages,
                status
        );
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeInt(identities);
        parcel.writeInt(messages);
        parcel.writeInt(status.ordinal());
    }

}

File: scatterbrainSDK/src/main/java/net/ballmerlabs/scatterbrainsdk/internal/HandshakeResult.kt
package net.ballmerlabs.scatterbrainsdk.internal

import android.os.Parcel
import android.os.Parcelable

/**
 * Transaction statistics from a handshake with a Scatterbrain peer.
 *
 * @property identities number of identities transferred
 * @property messages number of messages transferred
 * @property stackTrace error status of the transaction
 * @property success true if the transaction completed successfully
 */
open class HandshakeResult : Parcelable {
    enum class TransactionStatus {
        STATUS_SUCCESS, STATUS_FAIL
    }

    protected constructor(parcel: Parcel) {
        identities = parcel.readInt()
        messages = parcel.readInt()
        status = TransactionStatus.values()[parcel.readInt()]
    }

    val success: Boolean
        get() = status == TransactionStatus.STATUS_SUCCESS
    val identities: Int
    val messages: Int
    val status: TransactionStatus

    constructor(
            identities: Int,
            messages: Int,
            status: TransactionStatus
    ) {
        this.status = status
        this.messages = messages
        this.identities = identities
    }

    fun from(stats: HandshakeResult): HandshakeResult {
        val status: TransactionStatus = if (stats.status == TransactionStatus.STATUS_FAIL ||
                this.status == TransactionStatus.STATUS_FAIL) {
            TransactionStatus.STATUS_FAIL
        } else {
            stats.status
        }
        return HandshakeResult(
                stats.identities + identities,
                stats.messages + messages,
                status
        )
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(parcel: Parcel, i: Int) {
        parcel.writeInt(identities)
        parcel.writeInt(messages)
        parcel.writeInt(status.ordinal)
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<HandshakeResult> = object : Parcelable.Creator<HandshakeResult> {
            override fun createFromParcel(parcel: Parcel): HandshakeResult {
                return HandshakeResult(parcel)
            }

            override fun newArray(size: Int): Array<HandshakeResult?> {
                return arrayOfNulls(size)
            }
        }
    }
}